### PR TITLE
clean CP implementation for flash attention and cuDNN 9.6

### PR DESCRIPTION
**Pull request by xrennvidia**
# Description

- Flash attention: only use varlen kernels for THD format
- cuDNN >=9.6: fix implementation for packed format of softmax_lse, enabled THD+GQA, enabled SWA with KV all_gather communications.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refractor

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -8151,10 +8169,10 @@ def forward(
 
             pad_between_seqs = (
                 cu_seqlens_q_padded is not None
-                and not torch.equal(cu_seqlens_q_padded, cu_seqlens_q)
+                and not torch.equal(cu_seqlens_q_padded[:-1], cu_seqlens_q[:-1])
             ) or (
                 cu_seqlens_kv_padded is not None
-                and not torch.equal(cu_seqlens_kv_padded, cu_seqlens_kv)
+                and not torch.equal(cu_seqlens_kv_padded[:-1], cu_seqlens_kv[:-1])
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention.py:8175]: I notice that there are a couple of places you switched from the whole tensor to `tensor[:-1]`. Could the last element be different? Is the `cu_seqlens_q/kv_padded[-1]` somewhat special?

**Review by cyanguwa — APPROVED** [review]
LGTM

File: transformer_engine/pytorch/attention.py (hunk 1 above)
Comment (xrennvidia) [reply transformer_engine/pytorch/attention.py:8175]: The last element either mean the number of padding tokens at the end of packed sequence, or the length of last sequence segment that are packed. I think they are not necessary to be considered while we check if there are padding between sequence segments, because there are no valid tokens after them. Do you have any concern about this?

**Review by Trile9502 — APPROVED** [review]
